db/wal_edit.cc

```diff
@@ -0,0 +1,175 @@
+        }
+        metadata_.SetSizeInBytes(size);
+        break;
+      }
+      // TODO: process future tags such as checksum.
+      case WalAdditionTag::kTerminate:
+        return Status::OK();
+      default: {
+        std::stringstream ss;
+        ss << "Unknown tag " << tag_value;
+        return Status::Corruption(class_name, ss.str());
+      }
+    }
+  }
+}
+
+JSONWriter& operator<<(JSONWriter& jw, const WalAddition& wal) {
+  jw << "LogNumber" << wal.GetLogNumber() << "SizeInBytes"
+     << wal.GetMetadata().GetSizeInBytes();
+  return jw;
+}
+
+std::ostream& operator<<(std::ostream& os, const WalAddition& wal) {
+  os << "log_number: " << wal.GetLogNumber()
+     << " size_in_bytes: " << wal.GetMetadata().GetSizeInBytes();
+  return os;
+}
+
+std::string WalAddition::DebugString() const {
+  std::ostringstream oss;
+  oss << *this;
+  return oss.str();
+}
+
+void WalDeletion::EncodeTo(std::string* dst) const {
+  PutVarint64(dst, number_);
+}
+
+Status WalDeletion::DecodeFrom(Slice* src) {
+  constexpr char class_name[] = "WalDeletion";
+
+  if (!GetVarint64(src, &number_)) {
+    return Status::Corruption(class_name, "Error decoding WAL log number");
+  }
+
+  return Status::OK();
+}
+
+JSONWriter& operator<<(JSONWriter& jw, const WalDeletion& wal) {
+  jw << "LogNumber" << wal.GetLogNumber();
+  return jw;
+}
+
+std::ostream& operator<<(std::ostream& os, const WalDeletion& wal) {
+  os << "log_number: " << wal.GetLogNumber();
+  return os;
+}
+
+std::string WalDeletion::DebugString() const {
+  std::ostringstream oss;
+  oss << *this;
+  return oss.str();
+}
+
+Status WalSet::AddWal(const WalAddition& wal) {
+  auto it = wals_.lower_bound(wal.GetLogNumber());
+  if (wal.GetMetadata().HasSize()) {
+    // The WAL must exist without size.
+    if (it == wals_.end() || it->first != wal.GetLogNumber()) {
+      std::stringstream ss;
+      ss << "WAL " << wal.GetLogNumber() << " is not created before closing";
+      return Status::Corruption("WalSet", ss.str());
+    }
+    if (it->second.HasSize()) {
+      std::stringstream ss;
+      ss << "WAL " << wal.GetLogNumber() << " is closed more than once";
+      return Status::Corruption("WalSet", ss.str());
+    }
+    it->second = wal.GetMetadata();
+  } else {
+    // The WAL must not exist beforehand.
+    if (it != wals_.end() && it->first == wal.GetLogNumber()) {
+      std::stringstream ss;
+      ss << "WAL " << wal.GetLogNumber() << " is created more than once";
+      return Status::Corruption("WalSet", ss.str());
+    }
+    wals_[wal.GetLogNumber()] = wal.GetMetadata();
+  }
+  return Status::OK();
+}
+
+Status WalSet::AddWals(const WalAdditions& wals) {
+  Status s;
+  for (const WalAddition& wal : wals) {
+    s = AddWal(wal);
+    if (!s.ok()) {
+      break;
+    }
+  }
+  return s;
+}
+
+Status WalSet::DeleteWal(const WalDeletion& wal) {
+  auto it = wals_.lower_bound(wal.GetLogNumber());
```

Comment:
`find` is sufficient here, i.e. no need for `lower_bound` since we're not inserting.